### PR TITLE
Post about caching middleware

### PR DESCRIPTION
**Pull request by ezimuel**
A new post on how to build a caching middleware using Expressive.

### CODE REVIEW

**Comment by ezimuel** [comment]
@akrabat, @froschdesign can you review this new post blog? Thanks!

File: data/posts/2017-04-18-caching-middleware.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+memory, using [memcached](https://memcached.org)) and use it for next requests.
+In this way we can bypass the execution of the previous middlewares starting
+from the second request.
+
+Of course, this technique can only be applied for static contents, that do not
+require update for each HTTP request.
+
+## Implement a caching middleware
+
+Imagine we want to create a simple cache system with [Expressive](https://docs.zendframework.com/zend-expressive/)
```
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: Full stop is missing.

File: data/posts/2017-04-18-caching-middleware.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+```
+
+In this example, we used the [PSR-15](https://github.com/php-fig/fig-standards/blob/master/proposed/http-middleware/middleware.md)
+proposal to implement the Middleware interface using the `process()` function.
+This is the suggested way to implement middleware in Expressive 2.0.
+
+The idea of this middleware is quite simple, if the caching system is enabled
+and if the requested URL matches an existing cache file, we return the cache
+content as [HtmlResponse](https://github.com/zendframework/zend-diactoros/blob/master/src/Response/HtmlResponse.php)
+, ending the execution flow.
```
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: New line creates an extra whitespace.

File: data/posts/2017-04-18-caching-middleware.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+$app->get('/about', [
+    Action\CacheMiddleware::class,
+    Action\AboutAction::class
+], 'about');
+```
+
+The middleware actions to be excuted for the /about URL are `CacheMiddleware`
+and `AboutAction` in this order. The `$app` object is the instance of `Zend\Expressive\Application`
+, the main class that manages the execution of an Expressive application.
```
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: New line creates an extra whitespace.

File: data/posts/2017-04-18-caching-middleware.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+## How enable the cache for specific routes
+
+We mentioned early that this caching mechanism works fine for static contents.
+That means we need a way to enable the cache only for specific routes.
+
+We can simply add the `CacheMiddleware` class as first middleware to be executed
+for all the routes representing static contents.
+
+For instance, imagine to have a `/about` route that show an about page of your
+web site. We can add the `CacheMiddleware as follow:
```
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: Backtick is missing after "CacheMiddleware".

File: data/posts/2017-04-18-caching-middleware.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+you can omit the content putting a `.gitignore` file inside the `cache` folder
+with the following content:
+
+```php
+*
+!.gitignore
+```
+
+Finally, in order to activate the caching system we need to add the
+`CacheMiddleware` class as service. In our example, we used [zend-servicamanager](https://github.com/zendframework/zend-servicemanager)
```
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: Typing error in "servicemanager".

**Comment by froschdesign** [comment]
@ezimuel 
For all code examples, I would prefer it if the ZF coding standard would be used.

File: data/posts/2017-04-18-caching-middleware.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+use Interop\Http\ServerMiddleware\DelegateInterface;
+use Interop\Http\ServerMiddleware\MiddlewareInterface as ServerMiddlewareInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Diactoros\Response\HtmlResponse;
+
+class CacheMiddleware implements ServerMiddlewareInterface
+{
+    protected $config;
+
+    public function __construct($config)
```
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: Typehint for `$config`?

File: data/posts/2017-04-18-caching-middleware.md (hunk 2 above)
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: Should we link to the documentation?

https://zendframework.github.io/zend-diactoros/custom-responses/#html-responses

**Comment by froschdesign** [comment]
@ezimuel 
I like the simpleness of the caching solution in your article and the focus on the middleware principle. I see the possibility of another blog post with zend-cache.

**Comment by ezimuel** [comment]
@froschdesign thanks for your review and positive comment. You right, this caching system can be improved using zend-cache. Do you want to write a proposal for a new post?

File: data/posts/2017-04-18-caching-middleware.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+url_author: http://www.zimuel.it
+permalink: /blog/2017-04-18-caching-middleware.html
+categories:
+- blog
+- expressive
+
+---
+
+Performance is one of the key feature for web application. Using a middleware
+architecture become very simple to implement a caching system in PHP.
```
Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: s/become/makes it/

File: data/posts/2017-04-18-caching-middleware.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+The general idea is to store the response output of a URL in a file (or in
+memory, using [memcached](https://memcached.org)) and use it for next requests.
```
Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: s/next/subsequent/

File: data/posts/2017-04-18-caching-middleware.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+In this way we can bypass the execution of the previous middlewares starting
+from the second request.
+
+Of course, this technique can only be applied for static contents, that do not
```
Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: s/contents, that do/content, that does/

File: data/posts/2017-04-18-caching-middleware.md [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+        return $response;
+    }
+}
+```
+
+In this example, we used the [PSR-15](https://github.com/php-fig/fig-standards/blob/master/proposed/http-middleware/middleware.md)
+proposal to implement the Middleware interface using the `process()` function.
+This is the suggested way to implement middleware in Expressive 2.0.
+
+The idea of this middleware is quite simple, if the caching system is enabled
```
Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: /simple, if/simple. If/

File: data/posts/2017-04-18-caching-middleware.md [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+function of PHP to retrieve the modification file time.
+
+> Note: if you want to use memcached instead of file you need to replace the
+> `file_get_contents` and `file_put_contents` functions with `Memcached::get`
+> and `Memcached::set`. Moreover, you do not need to check for lifetime
+> because when you set a content in memcached you can specify the expiration
+> directly.
+
+In order to pass the `$config` dependency, we can use a simple factory class.
+Below is reported an example:
```
Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: /Below is reported/This is/

File: data/posts/2017-04-18-caching-middleware.md [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+                }
+            }
+        }
+        return new CacheMiddleware($config['cache']);
+    }
+}
+```
+Following the folder structure of Expressive, we can store this configuration in
+a simple PHP file in the config/autoload directory. For instance, we can store
+it in config/autoload/cache.local.php file, as follow:
```
Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: add backticks around the filename.

Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: s/follow/follows/

File: data/posts/2017-04-18-caching-middleware.md [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+        'factories' => [
+            App\Action\CacheMiddleware::class => App\Action\CacheFactory::class
+        ]
+    ]
+];
+```
+
+## How enable the cache for specific routes
+
+We mentioned early that this caching mechanism works fine for static contents.
```
Comment (akrabat) [inline data/posts/2017-04-18-caching-middleware.md]: s/contents/content/

**Comment by ezimuel** [comment]
@akrabat and @froschdesign thanks for your feedbacks!

File: data/posts/2017-04-18-caching-middleware.md (hunk 1 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 2 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: I changed with the documentation link, thanks.

File: data/posts/2017-04-18-caching-middleware.md (hunk 3 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 4 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 5 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 6 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 7 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 8 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 9 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md [hunk 14]
```diff
@@ -179,8 +179,8 @@ $app->get('/about', [
 ```
 
 The middleware actions to be excuted for the /about URL are `CacheMiddleware`
```
Comment (froschdesign) [inline data/posts/2017-04-18-caching-middleware.md]: Backticks are missing for `/about`-

File: data/posts/2017-04-18-caching-middleware.md (hunk 10 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 11 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 12 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 13 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

File: data/posts/2017-04-18-caching-middleware.md (hunk 14 above)
Comment (ezimuel) [inline data/posts/2017-04-18-caching-middleware.md]: Fixed

**Pull request merged**